File: 2020/Python/02-PasswordPhilosophy/PasswordPhilosophy.py
#!/usr/bin/env python3

# Password Philosophy
# https://adventofcode.com/2020/day/2

def read_file(name):
    file = open(name)
    return list(file.readlines())

class PasswordPolicy:
    def __init__(self, password, required_char, value1, value2):
        self.password = password
        self.required_char = required_char
        self.value1 = value1
        self.value2 = value2

    def is_valid_policy_one(self):
        count = len(list(filter(lambda char: char == self.required_char, self.password)))
        return count in range(self.value1, self.value2+1)

    def is_valid_policy_two(self):
        char1_valid = self.password[self.value1-1] == self.required_char
        char2_valid = self.password[self.value2-1] == self.required_char
        return char1_valid != char2_valid

def parse_password_policy(policy):
    components = policy.split()
    values = parse_values(components[0])
    required_char = components[1][0]
    password = components[2]
    return PasswordPolicy(password, required_char, values[0], values[1])

def parse_values(string):
    components = string.split("-")
    return (int(components[0]), int(components[1]))

password_policies = list(map(parse_password_policy, read_file('02-input.txt')))
valid_passwords_policy_one = list(filter(PasswordPolicy.is_valid_policy_one, password_policies))
print("There are {0} valid passwords for policy one".format(len(valid_passwords_policy_one)))

valid_passwords_policy_two = list(filter(PasswordPolicy.is_valid_policy_two, password_policies))
print("There are {0} valid passwords for policy two".format(len(valid_passwords_policy_two)))

File: 2020/Swift/02-PasswordPhilosophy/PasswordPhilosophy.swift
#!/usr/bin/swift

// Password Philosophy
// https://adventofcode.com/2020/day/2

import Foundation

func readFile(named name: String) -> [String] {
    let currentDirectoryURL = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
    let fileURL = URL(fileURLWithPath: name + ".txt", relativeTo: currentDirectoryURL)
    guard let content = try? String(contentsOf: fileURL, encoding: String.Encoding.utf8) else {
        print("Unable to read input file \(name)")
        print("Current directory: \(currentDirectoryURL)")
        return []
    }
    return content.components(separatedBy: .newlines)
}

extension StringProtocol {
    subscript(offset: Int) -> Character {
        self[index(startIndex, offsetBy: offset)]
    }
}

struct PasswordPolicy {
    let password: String
    let requiredChar: Character
    let integer1: UInt
    let integer2: UInt

    var minRequired: UInt { integer1 }
    var maxRequired: UInt { integer2 }

    var position1: Int { Int(integer1) }
    var position2: Int { Int(integer2) }

    // Policy one says that the required character must occur
    // between minRequired and maxRequired times in the string.
    func isValidForPolicyOne() -> Bool {
        let requiredCharCount = UInt(password.filter { $0 == requiredChar }.count)
        return minRequired...maxRequired ~= requiredCharCount
    }

    // Policy two says that the required character must be
    // in position1 or position2 but not in both.
    func isValidForPolicyTwo() -> Bool {
        let position1Valid = password[position1-1] == requiredChar
        let position2Valid = password[position2-1] == requiredChar
        return position1Valid != position2Valid
    }
}

func parsePasswordPolicy(string: String) -> PasswordPolicy? {
    let components = string.split(separator: " ")
    guard components.count == 3 else { return nil }

    guard let (integer1, integer2) = parseIntegers(string: components[0]),
          let requiredChar = parseRequiredChar(string: components[1]) else {
        return nil
    }
    let password = String(components[2])

    return PasswordPolicy(password: password,
                          requiredChar: requiredChar,
                          integer1: integer1,
                          integer2: integer2)
}

func parseIntegers(string: Substring) -> (UInt, UInt)? {
    let components = string.split(separator: "-")
    guard components.count == 2 else { return nil }
    guard let integer1 = UInt(components[0]) else { return nil }
    guard let integer2 = UInt(components[1]) else { return nil }
    return (integer1, integer2)
}

func parseRequiredChar(string: Substring) -> Character? {
    guard string.last == ":" else { return nil }
    return string.dropLast().first
}

let passwordPolicyStrings = readFile(named: "02-input")
let passwordPolicies = passwordPolicyStrings.compactMap { parsePasswordPolicy(string: $0) }

let validPasswordsPolicyOne = passwordPolicies.filter { $0.isValidForPolicyOne() }
print("There are \(validPasswordsPolicyOne.count) valid passwords for policy one")

let validPasswordsPolicyTwo = passwordPolicies.filter { $0.isValidForPolicyTwo() }
print("There are \(validPasswordsPolicyTwo.count) valid passwords for policy two")
